### ISSUE TITLE
chatglm2-6b-32k web_demo.py 多gpu部署bug

### ISSUE THREAD

**Issue by yang1111-gif** [THUDM/ChatGLM-6B#1364]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

![image](https://github.com/THUDM/ChatGLM-6B/assets/71382201/65b76d63-e2c0-4c91-a385-1e24af8305e8)
zh这是我在web_demo.py里面的模型加载代码，我这么运行，代码报错，如下：
![image](https://github.com/THUDM/ChatGLM-6B/assets/71382201/deaedc9b-9c15-453b-bb9a-9b6a72ce3492)
然后，w我加了device_map="auto"，还是报错，如下：
![image](https://github.com/THUDM/ChatGLM-6B/assets/71382201/50f85077-79d8-422c-869e-8db7cfa965fe)

所以，请问我该怎么修改，才能多卡部署成功，请指教，谢谢。
ps:模型已经下载在本地了。

### Expected Behavior

请指教

### Steps To Reproduce

请指教

### Environment

```markdown
- OS:
- Python:
- Transformers:
- PyTorch:
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) :
```

### Anything else?

请指教

**Comment by efwfe** [comment]
`def auto_configure_device_map(num_gpus: int) -> Dict[str, int]:
    # transformer.word_embeddings 占用1层
    # transformer.final_layernorm 和 lm_head 占用1层
    # transformer.layers 占用 28 层
    # 总共30层分配到num_gpus张卡上
    num_trans_layers = 28
    per_gpu_layers = 30 / num_gpus

    # bugfix: 在linux中调用torch.embedding传入的weight,input不在同一device上,导致RuntimeError
    # windows下 model.device 会被设置成 transformer.word_embeddings.device
    # linux下 model.device 会被设置成 lm_head.device
    # 在调用chat或者stream_chat时,input_ids会被放到model.device上
    # 如果transformer.word_embeddings.device和model.device不同,则会导致RuntimeError
    # 因此这里将transformer.word_embeddings,transformer.final_layernorm,lm_head都放到第一张卡上
    device_map = {'transformer.embedding.word_embeddings': 0,

'transformer.encoder.final_layernorm': 3,
'transformer.rotary_pos_emb':0,
'transformer.output_layer': 3,
'transformer.final_layernorm': 0, 'lm_head': 0}
    used = 2
    gpu_target = 0
    for i in range(num_trans_layers):
        if used >= per_gpu_layers:
            gpu_target += 1
            used = 0
        assert gpu_target < num_gpus
        device_map[f'transformer.encoder.layers.{i}'] = gpu_target
        used += 1

    return device_map`

修改utils.py,linux 4张GPU 修改后正常

**Comment by yang1111-gif** [comment]
> ```python
> def auto_configure_device_map(num_gpus: int) -> Dict[str, int]:
>     # transformer.word_embeddings 占用1层
>     # transformer.final_layernorm 和 lm_head 占用1层
>     # transformer.layers 占用 28 层
>     # 总共30层分配到num_gpus张卡上
>     num_trans_layers = 28
>     per_gpu_layers = 30 / num_gpus
> 
>     # bugfix: 在linux中调用torch.embedding传入的weight,input不在同一device上,导致RuntimeError
>     # windows下 model.device 会被设置成 transformer.word_embeddings.device
>     # linux下 model.device 会被设置成 lm_head.device
>     # 在调用chat或者stream_chat时,input_ids会被放到model.device上
>     # 如果transformer.word_embeddings.device和model.device不同,则会导致RuntimeError
>     # 因此这里将transformer.word_embeddings,transformer.final_layernorm,lm_head都放到第一张卡上
>     device_map = {'transformer.embedding.word_embeddings': 0,
> 
> 'transformer.encoder.final_layernorm': 3,
> 'transformer.rotary_pos_emb':0,
> 'transformer.output_layer': 3,
> 'transformer.final_layernorm': 0, 'lm_head': 0}
>     used = 2
>     gpu_target = 0
>     for i in range(num_trans_layers):
>         if used >= per_gpu_layers:
>             gpu_target += 1
>             used = 0
>         assert gpu_target < num_gpus
>         device_map[f'transformer.encoder.layers.{i}'] = gpu_target
>         used += 1
> 
>     return device_map
> ```
> 
> 修改utils.py,linux 4张GPU 修改后正常

您好，非常感谢您的回复和建议。您说的这个方法我试了，还是报错，如下所示：

web.demo.py的代码如下：
![Uploading image.png…]()

6b运行正常，就这32k不对。还请大佬指教

**Comment by Atarius0** [comment]
如果是modeling_chatglm.py中的
presents = torch.cat((presents, kv_cache), dim=0)
此行报错
我在前面加上了kv_cache = kv_cache.to(presents.device)
可以运行

**Comment by wxj127** [comment]
> ```python
> def auto_configure_device_map(num_gpus: int) -> Dict[str, int]:
>     # transformer.word_embeddings 占用1层
>     # transformer.final_layernorm 和 lm_head 占用1层
>     # transformer.layers 占用 28 层
>     # 总共30层分配到num_gpus张卡上
>     num_trans_layers = 28
>     per_gpu_layers = 30 / num_gpus
> 
>     # bugfix: 在linux中调用torch.embedding传入的weight,input不在同一device上,导致RuntimeError
>     # windows下 model.device 会被设置成 transformer.word_embeddings.device
>     # linux下 model.device 会被设置成 lm_head.device
>     # 在调用chat或者stream_chat时,input_ids会被放到model.device上
>     # 如果transformer.word_embeddings.device和model.device不同,则会导致RuntimeError
>     # 因此这里将transformer.word_embeddings,transformer.final_layernorm,lm_head都放到第一张卡上
>     device_map = {'transformer.embedding.word_embeddings': 0,
> 
> 'transformer.encoder.final_layernorm': 3,
> 'transformer.rotary_pos_emb':0,
> 'transformer.output_layer': 3,
> 'transformer.final_layernorm': 0, 'lm_head': 0}
>     used = 2
>     gpu_target = 0
>     for i in range(num_trans_layers):
>         if used >= per_gpu_layers:
>             gpu_target += 1
>             used = 0
>         assert gpu_target < num_gpus
>         device_map[f'transformer.encoder.layers.{i}'] = gpu_target
>         used += 1
> 
>     return device_map
> ```
> 
> 修改utils.py,linux 4张GPU 修改后正常

试了这个方法 没有用，我也是ubantu系统

**Comment by wxj127** [comment]
> kv_cache = kv_cache.to(presents.device)
![image](https://github.com/THUDM/ChatGLM-6B/assets/65206810/1991fb3e-c670-448c-a16c-0074596dffd7)
这样吗？

**Comment by wxj127** [comment]
> 如果是modeling_chatglm.py中的 presents = torch.cat((presents, kv_cache), dim=0) 此行报错 我在前面加上了kv_cache = kv_cache.to(presents.device) 多卡部署就成功了

你这个有用 ，成功了

**Comment by 300id** [comment]
哭了，用了你这个还是报一样的错。还有别的办法吗


> > 如果是modeling_chatglm.py中的 presents = torch.cat((presents, kv_cache), dim=0) 此行报错 我在前面加上了kv_cache = kv_cache.to(presents.device) 多卡部署就成功了
> 
> 你这个有用 ，成功了
